File: common/src/main/java/dev/mayaqq/estrogen/registry/EstrogenFluids.java
package dev.mayaqq.estrogen.registry;

import dev.mayaqq.estrogen.Estrogen;
import dev.mayaqq.estrogen.registry.blocks.fluids.EstrogenLiquidBlock;
import dev.mayaqq.estrogen.utils.EstrogenColors;
import dev.mayaqq.estrogen.utils.registry.EstrogenFluidEntry;
import dev.mayaqq.estrogen.utils.registry.EstrogenFluidRegistrar;
import earth.terrarium.botarium.common.registry.fluid.BotariumFlowingFluid;
import earth.terrarium.botarium.common.registry.fluid.BotariumSourceFluid;
import earth.terrarium.botarium.common.registry.fluid.FluidBucketItem;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.Rarity;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.material.MapColor;

public class EstrogenFluids {

    public static final EstrogenFluidRegistrar FLUIDS = EstrogenFluidRegistrar.create(Estrogen.REGISTRIES);


    // Setting all properties manually here
    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> LIQUID_ESTROGEN =
        FLUIDS.entry("liquid_estrogen", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .properties(p -> p.still(Estrogen.id("block/liquid_estrogen/liquid_estrogen_still"))
                .flowing(Estrogen.id("block/liquid_estrogen/liquid_estrogen_flow"))
                .overlay(Estrogen.id("block/liquid_estrogen/liquid_estrogen_flow"))
                .screenOverlay(new ResourceLocation("textures/misc/underwater.png"))
                .canConvertToSource(false)
                .canDrown(true)
                .canExtinguish(true)
                .canHydrate(true)
                .canPushEntity(true)
                .canSwim(true)
                .viscosity(1500)
                .density(1500))
            .renderType(() -> RenderType::translucent)
            .block(EstrogenLiquidBlock::new)
            .copyProperties(() -> Blocks.WATER)
            .properties(p -> p.mapColor(MapColor.COLOR_CYAN))
            .build()
            .bucket(FluidBucketItem::new)
            .properties(p -> p.craftRemainder(Items.BUCKET)
                .stacksTo(1)
                .rarity(Rarity.RARE))
            .build()
            .register();

    // Simple fluids done with transformers
    // Lava like
    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> MOLTEN_SLIME =
        FLUIDS.entry("molten_slime", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.lavaLikeFluid(MapColor.COLOR_GREEN, EstrogenColors.MOLTEN_SLIME.value))
            .transform(Transgenders.simpleBucket())
            .register();

    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> MOLTEN_AMETHYST =
        FLUIDS.entry("molten_amethyst", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.lavaLikeFluid(MapColor.COLOR_PURPLE, EstrogenColors.MOLTEN_AMETHYST.value))
            .transform(Transgenders.simpleBucket())
            .register();

    // Wate like
    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> TESTOSTERONE_MIXTURE =
        FLUIDS.entry("testosterone_mixture", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.waterLikeFluid(MapColor.TERRACOTTA_YELLOW, EstrogenColors.TESTOSTERONE_MIXTURE.value))
            .transform(Transgenders.simpleBucket())
            .register();

    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> FILTRATED_HORSE_URINE =
        FLUIDS.entry("filtrated_horse_urine", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.waterLikeFluid(MapColor.TERRACOTTA_YELLOW, EstrogenColors.FILTRATED_HORSE_URINE.value))
            .transform(Transgenders.simpleBucket())
            .register();

    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> HORSE_URINE =
        FLUIDS.entry("horse_urine", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.waterLikeFluid(MapColor.COLOR_YELLOW, EstrogenColors.HORSE_URINE.value))
            .transform(Transgenders.simpleBucket())
            .register();

    public static final EstrogenFluidEntry<BotariumSourceFluid, BotariumFlowingFluid> GENDER_FLUID =
        FLUIDS.entry("gender_fluid", BotariumSourceFluid::new, BotariumFlowingFluid::new)
            .transform(Transgenders.waterLikeFluid(MapColor.COLOR_CYAN, EstrogenColors.FILTRATED_HORSE_URINE.value))
            .transform(Transgenders.simpleBucket())
            .register();


}

File: src/main/kotlin/dev/mayaqq/estrogen/content/EstrogenFluids.kt
@file:Suppress("UNCHECKED_CAST")

package dev.mayaqq.estrogen.content

import com.teamresourceful.resourcefullib.client.fluid.registry.ResourcefulClientFluidRegistry
import com.teamresourceful.resourcefullib.common.fluid.ResourcefulBucketItem
import com.teamresourceful.resourcefullib.common.fluid.ResourcefulFlowingFluid
import com.teamresourceful.resourcefullib.common.fluid.registry.ResourcefulFluidRegistry
import com.teamresourceful.resourcefullib.common.registry.ResourcefulRegistries
import com.teamresourceful.resourcefullib.common.registry.ResourcefulRegistryType
import dev.mayaqq.estrogen.MOD_ID
import dev.mayaqq.estrogen.content.blocks.fluid.BaseEstrogenLiquidBlock
import dev.mayaqq.estrogen.content.blocks.fluid.EstrogenLiquidBlock
import dev.mayaqq.estrogen.content.fluids.registry.FluidRegistryProvider
import dev.mayaqq.estrogen.content.fluids.registry.fluid
import dev.mayaqq.estrogen.id
import dev.mayaqq.estrogen.mcid
import dev.mayaqq.estrogen.utils.EstrogenColors
import invoke.kitty.kritter.registry.api.Registrar
import net.minecraft.client.renderer.RenderType
import net.minecraft.core.registries.Registries
import net.minecraft.tags.FluidTags
import net.minecraft.world.item.Items
import net.minecraft.world.item.Rarity
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.material.Fluid
import net.minecraft.world.level.material.Fluids
import net.minecraft.world.level.material.MapColor

@Suppress("UnstableApiUsage")
object EstrogenFluids : Registrar<Fluid> by Registrar(MOD_ID, Registries.FLUID), FluidRegistryProvider {

    override val fluidRegistry: ResourcefulFluidRegistry = ResourcefulRegistries.create(ResourcefulRegistryType.FLUID, MOD_ID)
    override val clientFluidRegistry: ResourcefulClientFluidRegistry = ResourcefulClientFluidRegistry(MOD_ID)

    val LiquidEstrogen = fluid("liquid_estrogen", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
            properties {
                still(id("block/liquid_estrogen/liquid_estrogen_still"))
                flowing(id("block/liquid_estrogen/liquid_estrogen_flow"))
                screenOverlay(mcid("textures/misc/underwater.png"))
                canConvertToSource(false)
                canDrown(true)
                canExtinguish(true)
                canPushEntity(true)
                canSwim(true)
                viscosity(1500)
                density(1500)
                canHydrate(false)
            }
            renderType { RenderType.translucent() }
            block({ data, properties -> EstrogenLiquidBlock(data, properties,
                arrayOf(BaseEstrogenLiquidBlock.FluidInteraction { pos, state, fluidState ->
                    if (fluidState.`is`(Fluids.WATER)) {
                        return@FluidInteraction Blocks.CALCITE.defaultBlockState()
                    }
                    return@FluidInteraction null
                })
            )}) {
                initialPropertiesFrom(Blocks::WATER)
                properties {
                    mapColor(MapColor.COLOR_CYAN)
                }
            }
            bucket(::ResourcefulBucketItem) {
                properties {
                    craftRemainder(Items.BUCKET)
                    stacksTo(1)
                    rarity(Rarity.RARE)
                }
            }
        }
    val MoltenSlime = fluid("molten_slime", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        lavaLike(MapColor.COLOR_LIGHT_GREEN, EstrogenColors.MOLTEN_SLIME.toInt())
        simpleBucket()
    }
    val MoltenAmethyst = fluid("molten_amethyst", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        lavaLike(MapColor.COLOR_PURPLE, EstrogenColors.MOLTEN_AMETHYST.toInt())
        simpleBucket()
    }
    val TestosteroneMixture = fluid("testosterone_mixture", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        waterLike(MapColor.TERRACOTTA_YELLOW, EstrogenColors.TESTOSTERONE_MIXTURE.toInt())
        simpleBucket()
    }
    val FiltratedHorseUrine = fluid("filtrated_horse_urine", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        waterLike(MapColor.TERRACOTTA_YELLOW, EstrogenColors.FILTRATED_HORSE_URINE.toInt())
        simpleBucket()
    }
    val HorseUrine = fluid("horse_urine", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        waterLike(MapColor.COLOR_YELLOW, EstrogenColors.HORSE_URINE.toInt())
        simpleBucket()
    }
    val GenderFluid = fluid("gender_fluid", ResourcefulFlowingFluid::Still, ResourcefulFlowingFluid::Flowing) {
        properties {
            still(id("block/gender_fluid/gender_fluid_still"))
            flowing(id("block/gender_fluid/gender_fluid_flow"))
            overlay(id("block/gender_fluid/gender_fluid_flow"))
            screenOverlay(mcid("textures/misc/underwater.png"))
            canConvertToSource(false)
            canDrown(true)
            canExtinguish(true)
            canHydrate(false)
            canSwim(true)
            viscosity(1500)
            density(1500)
        }
        renderType { RenderType.translucent() }
        block({ data, properties -> EstrogenLiquidBlock(data, properties,
            arrayOf(BaseEstrogenLiquidBlock.FluidInteraction { pos, state, fluidState ->
                if (fluidState.`is`(FluidTags.LAVA)) {
                    return@FluidInteraction Blocks.PRISMARINE.defaultBlockState()
                }
                return@FluidInteraction null
            })
        )}) {
            initialPropertiesFrom(Blocks::WATER)
            properties {
                mapColor(MapColor.COLOR_PURPLE)
                emissiveRendering(Always)
            }
        }
        bucket(::ResourcefulBucketItem) {
            properties {
                craftRemainder(Items.BUCKET)
                stacksTo(1)
                rarity(Rarity.EPIC)
            }
        }
    }
}